From ff18dbaa05358bc615795c868c147c621aa6d8ac Mon Sep 17 00:00:00 2001
From: Michael Catanzaro <mcatanzaro@gnome.org>
Date: Mon, 24 Aug 2026 10:41:46 -0500
Subject: [PATCH] secret-unlock: fix crash when prompt is dismissed

Each time we process a new prompt from the queue, we assert to ensure
that we are not already processing a current prompt. Good idea, but the
state tracking is messed up.
---
 daemon/dbus/gkd-secret-unlock.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/daemon/dbus/gkd-secret-unlock.c b/daemon/dbus/gkd-secret-unlock.c
index a1a551f6..8681a117 100644
--- a/daemon/dbus/gkd-secret-unlock.c
+++ b/daemon/dbus/gkd-secret-unlock.c
@@ -155,6 +155,9 @@ mark_as_complete (GkdSecretUnlock *self, gboolean dismissed)
 		return FALSE;
 	self->completed = TRUE;
 
+	g_free (self->current);
+	self->current = NULL;
+
 	g_variant_builder_init (&builder, G_VARIANT_TYPE ("ao"));
 	for (i = 0; i < self->results->len; ++i) {
 		value = g_array_index (self->results, gchar*, i);
@@ -208,8 +211,6 @@ on_unlock_complete (GObject *object, GAsyncResult *res, gpointer user_data)
 
 	/* The user cancelled the protected auth prompt */
 	} else if (g_error_matches (error, GCK_ERROR, CKR_PIN_INCORRECT)) {
-		g_free (self->current);
-		self->current = NULL;
 		mark_as_complete (self, TRUE);
 
 	/* The operation was cancelled via Dismiss call */
-- 
2.55.0


